File: leetcode/LongestValidParentheses/LongestValidParentheses.cpp
class Solution {
public:
	/*
	 * 一种特殊的动态规划，依靠栈来判断括弧；
	 */
    int longestValidParentheses(string s) {
        stack<int> stk;
        stk.push(-1);
        int maxL = 0;
        for(int i = 0; i < s.size(); ++i){
            int t = stk.top();
            if(t != -1 && s[i] == ')' && s[t] == '(')
            {
                stk.pop();
                //DP思想
                maxL = max(maxL, i - stk.top());
            }else{
                stk.push(i);
            }
        }
        return maxL;
    }
};

File: leetcode/python3/longest-valid-parentheses.py
#
# @lc app=leetcode id=32 lang=python3
#
# [32] Longest Valid Parentheses
#
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        """
        if not s:
            return 0
        ans = 0
        stack = [-1]
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()
                if stack:
                    ans = max(ans, i-stack[-1])
                else:
                    stack.append(i)
        return ans 
        """
        ans = 0
        dp = [0] * len(s)
        for i in range(1, len(s)):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = dp[i-2] + 2
                else:
                    if i-dp[i-1] >= 1 and s[i-dp[i-1]-1] == '(':
                        dp[i] = dp[i-1] + dp[i-dp[i-1]-2] + 2
                ans = max(ans, dp[i])
        return ans

